[DebugMode][3/N] run_graph() method for fx.Node logging

pianpwk opened the pull request:
Allows logging on node calls, composed with existing `__torch_dispatch__` calls. Does this by creating a custom fx.Interpreter class with logging on `run_node()`.

Example from `debug_mode = DebugMode(); debug_mode.run_graph(gm, inputs)`:
```
  [node] l_flat_args_0_: placeholder
  [node] l__self____export_root_mlp_0_net1_weight: get_attr[L__self___mlp_0_net1_weight]()
  [node] l__self____export_root_mlp_0_net1_bias: get_attr[L__self___mlp_0_net1_bias]()
  [node] linear: call_function[torch._C._nn.linear](l_flat_args_0_, l__self____export_root_mlp_0_net1_weight, l__self____export_root_mlp_0_net1_bias)
    aten::t(dt: f32[16, 10]| S(0))
      aten::t(t: f32[4, 10])
    aten::addmm(dt: f32[16]| S(0), dt: f32[20, 10]| R, dt: f32[10, 16]| S(1))
      aten::addmm(t: f32[4], t: f32[20, 10], t: f32[10, 4])
  [node] outputs: call_function[torch._dynamo.variables.tensor.prim_redistribute](linear)
  [node] hook_result: call_function[torch._dynamo.variables.tensor.prim_to_local](outputs)
      aten::view(t: f32[20, 4], [20, 4])
  [node] input_tensor: call_function[torch.nn.functional.relu](hook_result, inplace=False)
      aten::relu(t: f32[20, 4])
  [node] input_tensor_1: call_function[torch._dynamo.variables.torch.prim from_local](input_tensor)
      aten::view(t: f32[20, 4], [20, 4])
  [node] l__self____export_root_mlp_0_net2_weight: get_attr[L__self___mlp_0_net2_weight]()
  [node] l__self____export_root_mlp_0_net2_bias: get_attr[L__self___mlp_0_net2_bias]()
  [node] linear_1: call_function[torch._C._nn.linear](input_tensor_1, l__self____export_root_mlp_0_net2_weight, l__self____export_root_mlp_0_net2_bias)
    aten::t(dt: f32[10, 16]| S(1))
      aten::t(t: f32[10, 4])
    aten::addmm(dt: f32[10]| R, dt: f32[20, 16]| S(1), dt: f32[16, 10]| S(0))
      redistribute_input(0, R -> P)
        redistribute_input(t: f32[10], trace: R->P)
          aten::div.Tensor(t: f32[10], 4)
      aten::addmm(t: f32[10], t: f32[20, 4], t: f32[4, 10])
  [node] outputs_1: call_function[torch._dynamo.variables.tensor.prim_redistribute](linear_1)
      redistribute_input(t: f32[20, 10], trace: P->R)
        _c10d_functional::all_reduce(t: f32[20, 10], sum, 5)
  [node] hook_result_1: call_function[torch._dynamo.variables.tensor.prim_to_local](outputs_1)
  [node] input_tensor_2: call_function[torch._dynamo.variables.torch.prim from_local](hook_result_1)
  [node] l__self____export_root_mlp_1_net1_weight: get_attr[L__self___mlp_1_net1_weight]()
  [node] l__self____export_root_mlp_1_net1_bias: get_attr[L__self___mlp_1_net1_bias]()
  [node] linear_2: call_function[torch._C._nn.linear](input_tensor_2, l__self____export_root_mlp_1_net1_weight, l__self____export_root_mlp_1_net1_bias)
    aten::t(dt: f32[16, 10]| S(0))
      aten::t(t: f32[4, 10])
    aten::addmm(dt: f32[16]| S(0), dt: f32[20, 10]| R, dt: f32[10, 16]| S(1))
  [node] outputs_2: call_function[torch._dynamo.variables.tensor.prim_redistribute](linear_2)
  [node] hook_result_2: call_function[torch._dynamo.variables.tensor.prim_to_local](outputs_2)
      aten::view(t: f32[20, 4], [20, 4])
  [node] input_tensor_3: call_function[torch.nn.functional.relu](hook_result_2, inplace=False)
      aten::relu(t: f32[20, 4])
  [node] input_tensor_4: call_function[torch._dynamo.variables.torch.prim from_local](input_tensor_3)
      aten::view(t: f32[20, 4], [20, 4])
  [node] l__self____export_root_mlp_1_net2_weight: get_attr[L__self___mlp_1_net2_weight]()
  [node] l__self____export_root_mlp_1_net2_bias: get_attr[L__self___mlp_1_net2_bias]()
  [node] linear_3: call_function[torch._C._nn.linear](input_tensor_4, l__self____export_root_mlp_1_net2_weight, l__self____export_root_mlp_1_net2_bias)
    aten::t(dt: f32[10, 16]| S(1))
      aten::t(t: f32[10, 4])
    aten::addmm(dt: f32[10]| R, dt: f32[20, 16]| S(1), dt: f32[16, 10]| S(0))
      redistribute_input(0, R -> P)
        redistribute_input(t: f32[10], trace: R->P)
          aten::div.Tensor(t: f32[10], 4)
      aten::addmm(t: f32[10], t: f32[20, 4], t: f32[4, 10])
  [node] outputs_3: call_function[torch._dynamo.variables.tensor.prim_redistribute](linear_3)
      redistribute_input(t: f32[20, 10], trace: P->R)
        _c10d_functional::all_reduce(t: f32[20, 10], sum, 5)
  [node] hook_result_3: call_function[torch._dynamo.variables.tensor.prim_to_local](outputs_3)
  [node] output: output
```

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #165499
* #165462
* #165497
* __->__ #165377
* #165461
* #165498



cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @msaroufim @dcci

pianpwk committed:
add run_graph

[ghstack-poisoned]

pianpwk committed:
Update on "add run_graph"




cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

pianpwk committed:
Update on "[DebugMode] run_graph() method for logging fx.node calls"

Allows logging on node calls, composed with existing `__torch_dispatch__` calls. Does this by creating a custom fx.Interpreter class with logging on `run_node()`.

Example from `debug_mode = DebugMode(); debug_mode.run_graph(gm, inputs)`:
```
  [node] l_flat_args_0_: placeholder
  [node] l__self____export_root_mlp_0_net1_weight: get_attr[L__self___mlp_0_net1_weight]()
  [node] l__self____export_root_mlp_0_net1_bias: get_attr[L__self___mlp_0_net1_bias]()
  [node] linear: call_function[torch._C._nn.linear](l_flat_args_0_, l__self____export_root_mlp_0_net1_weight, l__self____export_root_mlp_0_net1_bias)
    aten::t(dt: f32[16, 10]| S(0))
      aten::t(t: f32[4, 10])
    aten::addmm(dt: f32[16]| S(0), dt: f32[20, 10]| R, dt: f32[10, 16]| S(1))
      aten::addmm(t: f32[4], t: f32[20, 10], t: f32[10, 4])
  [node] outputs: call_function[torch._dynamo.variables.tensor.prim_redistribute](linear)
  [node] hook_result: call_function[torch._dynamo.variables.tensor.prim_to_local](outputs)
      aten::view(t: f32[20, 4], [20, 4])
  [node] input_tensor: call_function[torch.nn.functional.relu](hook_result, inplace=False)
      aten::relu(t: f32[20, 4])
  [node] input_tensor_1: call_function[torch._dynamo.variables.torch.prim from_local](input_tensor)
      aten::view(t: f32[20, 4], [20, 4])
  [node] l__self____export_root_mlp_0_net2_weight: get_attr[L__self___mlp_0_net2_weight]()
  [node] l__self____export_root_mlp_0_net2_bias: get_attr[L__self___mlp_0_net2_bias]()
  [node] linear_1: call_function[torch._C._nn.linear](input_tensor_1, l__self____export_root_mlp_0_net2_weight, l__self____export_root_mlp_0_net2_bias)
    aten::t(dt: f32[10, 16]| S(1))
      aten::t(t: f32[10, 4])
    aten::addmm(dt: f32[10]| R, dt: f32[20, 16]| S(1), dt: f32[16, 10]| S(0))
      redistribute_input(0, R -> P)
        redistribute_input(t: f32[10], trace: R->P)
          aten::div.Tensor(t: f32[10], 4)
      aten::addmm(t: f32[10], t: f32[20, 4], t: f32[4, 10])
  [node] outputs_1: call_function[torch._dynamo.variables.tensor.prim_redistribute](linear_1)
      redistribute_input(t: f32[20, 10], trace: P->R)
        _c10d_functional::all_reduce(t: f32[20, 10], sum, 5)
  [node] hook_result_1: call_function[torch._dynamo.variables.tensor.prim_to_local](outputs_1)
  [node] input_tensor_2: call_function[torch._dynamo.variables.torch.prim from_local](hook_result_1)
  [node] l__self____export_root_mlp_1_net1_weight: get_attr[L__self___mlp_1_net1_weight]()
  [node] l__self____export_root_mlp_1_net1_bias: get_attr[L__self___mlp_1_net1_bias]()
  [node] linear_2: call_function[torch._C._nn.linear](input_tensor_2, l__self____export_root_mlp_1_net1_weight, l__self____export_root_mlp_1_net1_bias)
    aten::t(dt: f32[16, 10]| S(0))
      aten::t(t: f32[4, 10])
    aten::addmm(dt: f32[16]| S(0), dt: f32[20, 10]| R, dt: f32[10, 16]| S(1))
  [node] outputs_2: call_function[torch._dynamo.variables.tensor.prim_redistribute](linear_2)
  [node] hook_result_2: call_function[torch._dynamo.variables.tensor.prim_to_local](outputs_2)
      aten::view(t: f32[20, 4], [20, 4])
  [node] input_tensor_3: call_function[torch.nn.functional.relu](hook_result_2, inplace=False)
      aten::relu(t: f32[20, 4])
  [node] input_tensor_4: call_function[torch._dynamo.variables.torch.prim from_local](input_tensor_3)
      aten::view(t: f32[20, 4], [20, 4])
  [node] l__self____export_root_mlp_1_net2_weight: get_attr[L__self___mlp_1_net2_weight]()
  [node] l__self____export_root_mlp_1_net2_bias: get_attr[L__self___mlp_1_net2_bias]()
  [node] linear_3: call_function[torch._C._nn.linear](input_tensor_4, l__self____export_root_mlp_1_net2_weight, l__self____export_root_mlp_1_net2_bias)
    aten::t(dt: f32[10, 16]| S(1))
      aten::t(t: f32[10, 4])
    aten::addmm(dt: f32[10]| R, dt: f32[20, 16]| S(1), dt: f32[16, 10]| S(0))
      redistribute_input(0, R -> P)
        redistribute_input(t: f32[10], trace: R->P)
          aten::div.Tensor(t: f32[10], 4)
      aten::addmm(t: f32[10], t: f32[20, 4], t: f32[4, 10])
  [node] outputs_3: call_function[torch._dynamo.variables.tensor.prim_redistribute](linear_3)
      redistribute_input(t: f32[20, 10], trace: P->R)
        _c10d_functional::all_reduce(t: f32[20, 10], sum, 5)
  [node] hook_result_3: call_function[torch._dynamo.variables.tensor.prim_to_local](outputs_3)
  [node] output: output
```




cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

pianpwk committed:
Update on "[DebugMode] run_graph() method for logging fx.node calls"

Allows logging on node calls, composed with existing `__torch_dispatch__` calls. Does this by creating a custom fx.Interpreter class with logging on `run_node()`.

Example from `debug_mode = DebugMode(); debug_mode.run_graph(gm, inputs)`:
```
  [node] l_flat_args_0_: placeholder
  [node] l__self____export_root_mlp_0_net1_weight: get_attr[L__self___mlp_0_net1_weight]()
  [node] l__self____export_root_mlp_0_net1_bias: get_attr[L__self___mlp_0_net1_bias]()
  [node] linear: call_function[torch._C._nn.linear](l_flat_args_0_, l__self____export_root_mlp_0_net1_weight, l__self____export_root_mlp_0_net1_bias)
    aten::t(dt: f32[16, 10]| S(0))
      aten::t(t: f32[4, 10])
    aten::addmm(dt: f32[16]| S(0), dt: f32[20, 10]| R, dt: f32[10, 16]| S(1))
      aten::addmm(t: f32[4], t: f32[20, 10], t: f32[10, 4])
  [node] outputs: call_function[torch._dynamo.variables.tensor.prim_redistribute](linear)
  [node] hook_result: call_function[torch._dynamo.variables.tensor.prim_to_local](outputs)
      aten::view(t: f32[20, 4], [20, 4])
  [node] input_tensor: call_function[torch.nn.functional.relu](hook_result, inplace=False)
      aten::relu(t: f32[20, 4])
  [node] input_tensor_1: call_function[torch._dynamo.variables.torch.prim from_local](input_tensor)
      aten::view(t: f32[20, 4], [20, 4])
  [node] l__self____export_root_mlp_0_net2_weight: get_attr[L__self___mlp_0_net2_weight]()
  [node] l__self____export_root_mlp_0_net2_bias: get_attr[L__self___mlp_0_net2_bias]()
  [node] linear_1: call_function[torch._C._nn.linear](input_tensor_1, l__self____export_root_mlp_0_net2_weight, l__self____export_root_mlp_0_net2_bias)
    aten::t(dt: f32[10, 16]| S(1))
      aten::t(t: f32[10, 4])
    aten::addmm(dt: f32[10]| R, dt: f32[20, 16]| S(1), dt: f32[16, 10]| S(0))
      redistribute_input(0, R -> P)
        redistribute_input(t: f32[10], trace: R->P)
          aten::div.Tensor(t: f32[10], 4)
      aten::addmm(t: f32[10], t: f32[20, 4], t: f32[4, 10])
  [node] outputs_1: call_function[torch._dynamo.variables.tensor.prim_redistribute](linear_1)
      redistribute_input(t: f32[20, 10], trace: P->R)
        _c10d_functional::all_reduce(t: f32[20, 10], sum, 5)
  [node] hook_result_1: call_function[torch._dynamo.variables.tensor.prim_to_local](outputs_1)
  [node] input_tensor_2: call_function[torch._dynamo.variables.torch.prim from_local](hook_result_1)
  [node] l__self____export_root_mlp_1_net1_weight: get_attr[L__self___mlp_1_net1_weight]()
  [node] l__self____export_root_mlp_1_net1_bias: get_attr[L__self___mlp_1_net1_bias]()
  [node] linear_2: call_function[torch._C._nn.linear](input_tensor_2, l__self____export_root_mlp_1_net1_weight, l__self____export_root_mlp_1_net1_bias)
    aten::t(dt: f32[16, 10]| S(0))
      aten::t(t: f32[4, 10])
    aten::addmm(dt: f32[16]| S(0), dt: f32[20, 10]| R, dt: f32[10, 16]| S(1))
  [node] outputs_2: call_function[torch._dynamo.variables.tensor.prim_redistribute](linear_2)
  [node] hook_result_2: call_function[torch._dynamo.variables.tensor.prim_to_local](outputs_2)
      aten::view(t: f32[20, 4], [20, 4])
  [node] input_tensor_3: call_function[torch.nn.functional.relu](hook_result_2, inplace=False)
      aten::relu(t: f32[20, 4])
  [node] input_tensor_4: call_function[torch._dynamo.variables.torch.prim from_local](input_tensor_3)
      aten::view(t: f32[20, 4], [20, 4])
  [node] l__self____export_root_mlp_1_net2_weight: get_attr[L__self___mlp_1_net2_weight]()
  [node] l__self____export_root_mlp_1_net2_bias: get_attr[L__self___mlp_1_net2_bias]()
  [node] linear_3: call_function[torch._C._nn.linear](input_tensor_4, l__self____export_root_mlp_1_net2_weight, l__self____export_root_mlp_1_net2_bias)
    aten::t(dt: f32[10, 16]| S(1))
      aten::t(t: f32[10, 4])
    aten::addmm(dt: f32[10]| R, dt: f32[20, 16]| S(1), dt: f32[16, 10]| S(0))
      redistribute_input(0, R -> P)
        redistribute_input(t: f32[10], trace: R->P)
          aten::div.Tensor(t: f32[10], 4)
      aten::addmm(t: f32[10], t: f32[20, 4], t: f32[4, 10])
  [node] outputs_3: call_function[torch._dynamo.variables.tensor.prim_redistribute](linear_3)
      redistribute_input(t: f32[20, 10], trace: P->R)
        _c10d_functional::all_reduce(t: f32[20, 10], sum, 5)
  [node] hook_result_3: call_function[torch._dynamo.variables.tensor.prim_to_local](outputs_3)
  [node] output: output
```




cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

pianpwk committed:
Update on "[DebugMode] run_graph() method for logging fx.node calls"

Allows logging on node calls, composed with existing `__torch_dispatch__` calls. Does this by creating a custom fx.Interpreter class with logging on `run_node()`.

Example from `debug_mode = DebugMode(); debug_mode.run_graph(gm, inputs)`:
```
  [node] l_flat_args_0_: placeholder
  [node] l__self____export_root_mlp_0_net1_weight: get_attr[L__self___mlp_0_net1_weight]()
  [node] l__self____export_root_mlp_0_net1_bias: get_attr[L__self___mlp_0_net1_bias]()
  [node] linear: call_function[torch._C._nn.linear](l_flat_args_0_, l__self____export_root_mlp_0_net1_weight, l__self____export_root_mlp_0_net1_bias)
    aten::t(dt: f32[16, 10]| S(0))
      aten::t(t: f32[4, 10])
    aten::addmm(dt: f32[16]| S(0), dt: f32[20, 10]| R, dt: f32[10, 16]| S(1))
      aten::addmm(t: f32[4], t: f32[20, 10], t: f32[10, 4])
  [node] outputs: call_function[torch._dynamo.variables.tensor.prim_redistribute](linear)
  [node] hook_result: call_function[torch._dynamo.variables.tensor.prim_to_local](outputs)
      aten::view(t: f32[20, 4], [20, 4])
  [node] input_tensor: call_function[torch.nn.functional.relu](hook_result, inplace=False)
      aten::relu(t: f32[20, 4])
  [node] input_tensor_1: call_function[torch._dynamo.variables.torch.prim from_local](input_tensor)
      aten::view(t: f32[20, 4], [20, 4])
  [node] l__self____export_root_mlp_0_net2_weight: get_attr[L__self___mlp_0_net2_weight]()
  [node] l__self____export_root_mlp_0_net2_bias: get_attr[L__self___mlp_0_net2_bias]()
  [node] linear_1: call_function[torch._C._nn.linear](input_tensor_1, l__self____export_root_mlp_0_net2_weight, l__self____export_root_mlp_0_net2_bias)
    aten::t(dt: f32[10, 16]| S(1))
      aten::t(t: f32[10, 4])
    aten::addmm(dt: f32[10]| R, dt: f32[20, 16]| S(1), dt: f32[16, 10]| S(0))
      redistribute_input(0, R -> P)
        redistribute_input(t: f32[10], trace: R->P)
          aten::div.Tensor(t: f32[10], 4)
      aten::addmm(t: f32[10], t: f32[20, 4], t: f32[4, 10])
  [node] outputs_1: call_function[torch._dynamo.variables.tensor.prim_redistribute](linear_1)
      redistribute_input(t: f32[20, 10], trace: P->R)
        _c10d_functional::all_reduce(t: f32[20, 10], sum, 5)
  [node] hook_result_1: call_function[torch._dynamo.variables.tensor.prim_to_local](outputs_1)
  [node] input_tensor_2: call_function[torch._dynamo.variables.torch.prim from_local](hook_result_1)
  [node] l__self____export_root_mlp_1_net1_weight: get_attr[L__self___mlp_1_net1_weight]()
  [node] l__self____export_root_mlp_1_net1_bias: get_attr[L__self___mlp_1_net1_bias]()
  [node] linear_2: call_function[torch._C._nn.linear](input_tensor_2, l__self____export_root_mlp_1_net1_weight, l__self____export_root_mlp_1_net1_bias)
    aten::t(dt: f32[16, 10]| S(0))
      aten::t(t: f32[4, 10])
    aten::addmm(dt: f32[16]| S(0), dt: f32[20, 10]| R, dt: f32[10, 16]| S(1))
  [node] outputs_2: call_function[torch._dynamo.variables.tensor.prim_redistribute](linear_2)
  [node] hook_result_2: call_function[torch._dynamo.variables.tensor.prim_to_local](outputs_2)
      aten::view(t: f32[20, 4], [20, 4])
  [node] input_tensor_3: call_function[torch.nn.functional.relu](hook_result_2, inplace=False)
      aten::relu(t: f32[20, 4])
  [node] input_tensor_4: call_function[torch._dynamo.variables.torch.prim from_local](input_tensor_3)
      aten::view(t: f32[20, 4], [20, 4])
  [node] l__self____export_root_mlp_1_net2_weight: get_attr[L__self___mlp_1_net2_weight]()
  [node] l__self____export_root_mlp_1_net2_bias: get_attr[L__self___mlp_1_net2_bias]()
  [node] linear_3: call_function[torch._C._nn.linear](input_tensor_4, l__self____export_root_mlp_1_net2_weight, l__self____export_root_mlp_1_net2_bias)
    aten::t(dt: f32[10, 16]| S(1))
      aten::t(t: f32[10, 4])
    aten::addmm(dt: f32[10]| R, dt: f32[20, 16]| S(1), dt: f32[16, 10]| S(0))
      redistribute_input(0, R -> P)
        redistribute_input(t: f32[10], trace: R->P)
          aten::div.Tensor(t: f32[10], 4)
      aten::addmm(t: f32[10], t: f32[20, 4], t: f32[4, 10])
  [node] outputs_3: call_function[torch._dynamo.variables.tensor.prim_redistribute](linear_3)
      redistribute_input(t: f32[20, 10], trace: P->R)
        _c10d_functional::all_reduce(t: f32[20, 10], sum, 5)
  [node] hook_result_3: call_function[torch._dynamo.variables.tensor.prim_to_local](outputs_3)
  [node] output: output
```




cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

SherlockNoMad commented on test/distributed/tensor/debug/test_debug_mode.py:
```diff
@@ -339,6 +353,103 @@ def f(x):
+  [node] outputs: call_function[torch._dynamo.variables.tensor.prim_redistribute](linear)
+  [node] hook_result: call_function[torch._dynamo.variables.tensor.prim_to_local](outputs)
+      aten::view(t: f32[20, 4], [20, 4])
+  [node] input_tensor: call_function[torch.nn.functional.relu](hook_result, inplace=False)
```
nit. 
input_tensor = call_function

SherlockNoMad commented on test/distributed/tensor/debug/test_debug_mode.py:
```diff
@@ -339,6 +353,103 @@ def f(x):
+            "mlp_1.net1": ColwiseParallel(),
+            "mlp_1.net2": RowwiseParallel(),
+        }
+        tp_model = parallelize_module(model, mesh["tp"], parallelize_plan)
```
seems duplicated with other tests, 
refactor to reuse.

SherlockNoMad reviewed (changes requested):
TBH, the use of this DebugMode.run_graph() would probably be niche. 

we will not directly expose dynamo produced gm to user.
Instead, we will expose post-dispatch aten fx graph. 

if we run DebugMode.run_graph() on post-dispatch aten fx graph, I suppose we would have "one node to one log" mapping. 

You can experiment with running aot_export_joint produced graph to see the actual effect.

pianpwk committed:
Update on "[DebugMode][3/N] run_graph() method for fx.Node logging"

Allows logging on node calls, composed with existing `__torch_dispatch__` calls. Does this by creating a custom fx.Interpreter class with logging on `run_node()`.

Example from `debug_mode = DebugMode(); debug_mode.run_graph(gm, inputs)`:
```
  [node] l_flat_args_0_: placeholder
  [node] l__self____export_root_mlp_0_net1_weight: get_attr[L__self___mlp_0_net1_weight]()
  [node] l__self____export_root_mlp_0_net1_bias: get_attr[L__self___mlp_0_net1_bias]()
  [node] linear: call_function[torch._C._nn.linear](l_flat_args_0_, l__self____export_root_mlp_0_net1_weight, l__self____export_root_mlp_0_net1_bias)
    aten::t(dt: f32[16, 10]| S(0))
      aten::t(t: f32[4, 10])
    aten::addmm(dt: f32[16]| S(0), dt: f32[20, 10]| R, dt: f32[10, 16]| S(1))
      aten::addmm(t: f32[4], t: f32[20, 10], t: f32[10, 4])
  [node] outputs: call_function[torch._dynamo.variables.tensor.prim_redistribute](linear)
  [node] hook_result: call_function[torch._dynamo.variables.tensor.prim_to_local](outputs)
      aten::view(t: f32[20, 4], [20, 4])
  [node] input_tensor: call_function[torch.nn.functional.relu](hook_result, inplace=False)
      aten::relu(t: f32[20, 4])
  [node] input_tensor_1: call_function[torch._dynamo.variables.torch.prim from_local](input_tensor)
      aten::view(t: f32[20, 4], [20, 4])
  [node] l__self____export_root_mlp_0_net2_weight: get_attr[L__self___mlp_0_net2_weight]()
  [node] l__self____export_root_mlp_0_net2_bias: get_attr[L__self___mlp_0_net2_bias]()
  [node] linear_1: call_function[torch._C._nn.linear](input_tensor_1, l__self____export_root_mlp_0_net2_weight, l__self____export_root_mlp_0_net2_bias)
    aten::t(dt: f32[10, 16]| S(1))
      aten::t(t: f32[10, 4])
    aten::addmm(dt: f32[10]| R, dt: f32[20, 16]| S(1), dt: f32[16, 10]| S(0))
      redistribute_input(0, R -> P)
        redistribute_input(t: f32[10], trace: R->P)
          aten::div.Tensor(t: f32[10], 4)
      aten::addmm(t: f32[10], t: f32[20, 4], t: f32[4, 10])
  [node] outputs_1: call_function[torch._dynamo.variables.tensor.prim_redistribute](linear_1)
      redistribute_input(t: f32[20, 10], trace: P->R)
        _c10d_functional::all_reduce(t: f32[20, 10], sum, 5)
  [node] hook_result_1: call_function[torch._dynamo.variables.tensor.prim_to_local](outputs_1)
  [node] input_tensor_2: call_function[torch._dynamo.variables.torch.prim from_local](hook_result_1)
  [node] l__self____export_root_mlp_1_net1_weight: get_attr[L__self___mlp_1_net1_weight]()
  [node] l__self____export_root_mlp_1_net1_bias: get_attr[L__self___mlp_1_net1_bias]()
  [node] linear_2: call_function[torch._C._nn.linear](input_tensor_2, l__self____export_root_mlp_1_net1_weight, l__self____export_root_mlp_1_net1_bias)
    aten::t(dt: f32[16, 10]| S(0))
      aten::t(t: f32[4, 10])
    aten::addmm(dt: f32[16]| S(0), dt: f32[20, 10]| R, dt: f32[10, 16]| S(1))
  [node] outputs_2: call_function[torch._dynamo.variables.tensor.prim_redistribute](linear_2)
  [node] hook_result_2: call_function[torch._dynamo.variables.tensor.prim_to_local](outputs_2)
      aten::view(t: f32[20, 4], [20, 4])
  [node] input_tensor_3: call_function[torch.nn.functional.relu](hook_result_2, inplace=False)
      aten::relu(t: f32[20, 4])
  [node] input_tensor_4: call_function[torch._dynamo.variables.torch.prim from_local](input_tensor_3)
      aten::view(t: f32[20, 4], [20, 4])
  [node] l__self____export_root_mlp_1_net2_weight: get_attr[L__self___mlp_1_net2_weight]()
  [node] l__self____export_root_mlp_1_net2_bias: get_attr[L__self___mlp_1_net2_bias]()
  [node] linear_3: call_function[torch._C._nn.linear](input_tensor_4, l__self____export_root_mlp_1_net2_weight, l__self____export_root_mlp_1_net2_bias)
    aten::t(dt: f32[10, 16]| S(1))
      aten::t(t: f32[10, 4])
    aten::addmm(dt: f32[10]| R, dt: f32[20, 16]| S(1), dt: f32[16, 10]| S(0))
      redistribute_input(0, R -> P)
        redistribute_input(t: f32[10], trace: R->P)
          aten::div.Tensor(t: f32[10], 4)
      aten::addmm(t: f32[10], t: f32[20, 4], t: f32[4, 10])
  [node] outputs_3: call_function[torch._dynamo.variables.tensor.prim_redistribute](linear_3)
      redistribute_input(t: f32[20, 10], trace: P->R)
        _c10d_functional::all_reduce(t: f32[20, 10], sum, 5)
  [node] hook_result_3: call_function[torch._dynamo.variables.tensor.prim_to_local](outputs_3)
  [node] output: output
```




cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

pianpwk committed:
Update on "[DebugMode][3/N] run_graph() method for fx.Node logging"

Allows logging on node calls, composed with existing `__torch_dispatch__` calls. Does this by creating a custom fx.Interpreter class with logging on `run_node()`.

Example from `debug_mode = DebugMode(); debug_mode.run_graph(gm, inputs)`:
```
  [node] l_flat_args_0_: placeholder
  [node] l__self____export_root_mlp_0_net1_weight: get_attr[L__self___mlp_0_net1_weight]()
  [node] l__self____export_root_mlp_0_net1_bias: get_attr[L__self___mlp_0_net1_bias]()
  [node] linear: call_function[torch._C._nn.linear](l_flat_args_0_, l__self____export_root_mlp_0_net1_weight, l__self____export_root_mlp_0_net1_bias)
    aten::t(dt: f32[16, 10]| S(0))
      aten::t(t: f32[4, 10])
    aten::addmm(dt: f32[16]| S(0), dt: f32[20, 10]| R, dt: f32[10, 16]| S(1))
      aten::addmm(t: f32[4], t: f32[20, 10], t: f32[10, 4])
  [node] outputs: call_function[torch._dynamo.variables.tensor.prim_redistribute](linear)
  [node] hook_result: call_function[torch._dynamo.variables.tensor.prim_to_local](outputs)
      aten::view(t: f32[20, 4], [20, 4])
  [node] input_tensor: call_function[torch.nn.functional.relu](hook_result, inplace=False)
      aten::relu(t: f32[20, 4])
  [node] input_tensor_1: call_function[torch._dynamo.variables.torch.prim from_local](input_tensor)
      aten::view(t: f32[20, 4], [20, 4])
  [node] l__self____export_root_mlp_0_net2_weight: get_attr[L__self___mlp_0_net2_weight]()
  [node] l__self____export_root_mlp_0_net2_bias: get_attr[L__self___mlp_0_net2_bias]()
  [node] linear_1: call_function[torch._C._nn.linear](input_tensor_1, l__self____export_root_mlp_0_net2_weight, l__self____export_root_mlp_0_net2_bias)
    aten::t(dt: f32[10, 16]| S(1))
      aten::t(t: f32[10, 4])
    aten::addmm(dt: f32[10]| R, dt: f32[20, 16]| S(1), dt: f32[16, 10]| S(0))
      redistribute_input(0, R -> P)
        redistribute_input(t: f32[10], trace: R->P)
          aten::div.Tensor(t: f32[10], 4)
      aten::addmm(t: f32[10], t: f32[20, 4], t: f32[4, 10])
  [node] outputs_1: call_function[torch._dynamo.variables.tensor.prim_redistribute](linear_1)
      redistribute_input(t: f32[20, 10], trace: P->R)
        _c10d_functional::all_reduce(t: f32[20, 10], sum, 5)
  [node] hook_result_1: call_function[torch._dynamo.variables.tensor.prim_to_local](outputs_1)
  [node] input_tensor_2: call_function[torch._dynamo.variables.torch.prim from_local](hook_result_1)
  [node] l__self____export_root_mlp_1_net1_weight: get_attr[L__self___mlp_1_net1_weight]()
  [node] l__self____export_root_mlp_1_net1_bias: get_attr[L__self___mlp_1_net1_bias]()
  [node] linear_2: call_function[torch._C._nn.linear](input_tensor_2, l__self____export_root_mlp_1_net1_weight, l__self____export_root_mlp_1_net1_bias)
    aten::t(dt: f32[16, 10]| S(0))
      aten::t(t: f32[4, 10])
    aten::addmm(dt: f32[16]| S(0), dt: f32[20, 10]| R, dt: f32[10, 16]| S(1))
  [node] outputs_2: call_function[torch._dynamo.variables.tensor.prim_redistribute](linear_2)
  [node] hook_result_2: call_function[torch._dynamo.variables.tensor.prim_to_local](outputs_2)
      aten::view(t: f32[20, 4], [20, 4])
  [node] input_tensor_3: call_function[torch.nn.functional.relu](hook_result_2, inplace=False)
      aten::relu(t: f32[20, 4])
  [node] input_tensor_4: call_function[torch._dynamo.variables.torch.prim from_local](input_tensor_3)
      aten::view(t: f32[20, 4], [20, 4])
  [node] l__self____export_root_mlp_1_net2_weight: get_attr[L__self___mlp_1_net2_weight]()
  [node] l__self____export_root_mlp_1_net2_bias: get_attr[L__self___mlp_1_net2_bias]()
  [node] linear_3: call_function[torch._C._nn.linear](input_tensor_4, l__self____export_root_mlp_1_net2_weight, l__self____export_root_mlp_1_net2_bias)
    aten::t(dt: f32[10, 16]| S(1))
      aten::t(t: f32[10, 4])
    aten::addmm(dt: f32[10]| R, dt: f32[20, 16]| S(1), dt: f32[16, 10]| S(0))
      redistribute_input(0, R -> P)
        redistribute_input(t: f32[10], trace: R->P)
          aten::div.Tensor(t: f32[10], 4)
      aten::addmm(t: f32[10], t: f32[20, 4], t: f32[4, 10])
  [node] outputs_3: call_function[torch._dynamo.variables.tensor.prim_redistribute](linear_3)
      redistribute_input(t: f32[20, 10], trace: P->R)
        _c10d_functional::all_reduce(t: f32[20, 10], sum, 5)
  [node] hook_result_3: call_function[torch._dynamo.variables.tensor.prim_to_local](outputs_3)
  [node] output: output
```




cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

pianpwk committed:
Update on "[DebugMode][3/N] run_graph() method for fx.Node logging"

Allows logging on node calls, composed with existing `__torch_dispatch__` calls. Does this by creating a custom fx.Interpreter class with logging on `run_node()`.

Example from `debug_mode = DebugMode(); debug_mode.run_graph(gm, inputs)`:
```
  [node] l_flat_args_0_: placeholder
  [node] l__self____export_root_mlp_0_net1_weight: get_attr[L__self___mlp_0_net1_weight]()
  [node] l__self____export_root_mlp_0_net1_bias: get_attr[L__self___mlp_0_net1_bias]()
  [node] linear: call_function[torch._C._nn.linear](l_flat_args_0_, l__self____export_root_mlp_0_net1_weight, l__self____export_root_mlp_0_net1_bias)
    aten::t(dt: f32[16, 10]| S(0))
      aten::t(t: f32[4, 10])
    aten::addmm(dt: f32[16]| S(0), dt: f32[20, 10]| R, dt: f32[10, 16]| S(1))
      aten::addmm(t: f32[4], t: f32[20, 10], t: f32[10, 4])
  [node] outputs: call_function[torch._dynamo.variables.tensor.prim_redistribute](linear)
  [node] hook_result: call_function[torch._dynamo.variables.tensor.prim_to_local](outputs)
      aten::view(t: f32[20, 4], [20, 4])
  [node] input_tensor: call_function[torch.nn.functional.relu](hook_result, inplace=False)
      aten::relu(t: f32[20, 4])
  [node] input_tensor_1: call_function[torch._dynamo.variables.torch.prim from_local](input_tensor)
      aten::view(t: f32[20, 4], [20, 4])
  [node] l__self____export_root_mlp_0_net2_weight: get_attr[L__self___mlp_0_net2_weight]()
  [node] l__self____export_root_mlp_0_net2_bias: get_attr[L__self___mlp_0_net2_bias]()
  [node] linear_1: call_function[torch._C._nn.linear](input_tensor_1, l__self____export_root_mlp_0_net2_weight, l__self____export_root_mlp_0_net2_bias)
    aten::t(dt: f32[10, 16]| S(1))
      aten::t(t: f32[10, 4])
    aten::addmm(dt: f32[10]| R, dt: f32[20, 16]| S(1), dt: f32[16, 10]| S(0))
      redistribute_input(0, R -> P)
        redistribute_input(t: f32[10], trace: R->P)
          aten::div.Tensor(t: f32[10], 4)
      aten::addmm(t: f32[10], t: f32[20, 4], t: f32[4, 10])
  [node] outputs_1: call_function[torch._dynamo.variables.tensor.prim_redistribute](linear_1)
      redistribute_input(t: f32[20, 10], trace: P->R)
        _c10d_functional::all_reduce(t: f32[20, 10], sum, 5)
  [node] hook_result_1: call_function[torch._dynamo.variables.tensor.prim_to_local](outputs_1)
  [node] input_tensor_2: call_function[torch._dynamo.variables.torch.prim from_local](hook_result_1)
  [node] l__self____export_root_mlp_1_net1_weight: get_attr[L__self___mlp_1_net1_weight]()
  [node] l__self____export_root_mlp_1_net1_bias: get_attr[L__self___mlp_1_net1_bias]()
  [node] linear_2: call_function[torch._C._nn.linear](input_tensor_2, l__self____export_root_mlp_1_net1_weight, l__self____export_root_mlp_1_net1_bias)
    aten::t(dt: f32[16, 10]| S(0))
      aten::t(t: f32[4, 10])
    aten::addmm(dt: f32[16]| S(0), dt: f32[20, 10]| R, dt: f32[10, 16]| S(1))
  [node] outputs_2: call_function[torch._dynamo.variables.tensor.prim_redistribute](linear_2)
  [node] hook_result_2: call_function[torch._dynamo.variables.tensor.prim_to_local](outputs_2)
      aten::view(t: f32[20, 4], [20, 4])
  [node] input_tensor_3: call_function[torch.nn.functional.relu](hook_result_2, inplace=False)
      aten::relu(t: f32[20, 4])
  [node] input_tensor_4: call_function[torch._dynamo.variables.torch.prim from_local](input_tensor_3)
      aten::view(t: f32[20, 4], [20, 4])
  [node] l__self____export_root_mlp_1_net2_weight: get_attr[L__self___mlp_1_net2_weight]()
  [node] l__self____export_root_mlp_1_net2_bias: get_attr[L__self___mlp_1_net2_bias]()
  [node] linear_3: call_function[torch._C._nn.linear](input_tensor_4, l__self____export_root_mlp_1_net2_weight, l__self____export_root_mlp_1_net2_bias)
    aten::t(dt: f32[10, 16]| S(1))
      aten::t(t: f32[10, 4])
    aten::addmm(dt: f32[10]| R, dt: f32[20, 16]| S(1), dt: f32[16, 10]| S(0))
      redistribute_input(0, R -> P)
        redistribute_input(t: f32[10], trace: R->P)
          aten::div.Tensor(t: f32[10], 4)
      aten::addmm(t: f32[10], t: f32[20, 4], t: f32[4, 10])
  [node] outputs_3: call_function[torch._dynamo.variables.tensor.prim_redistribute](linear_3)
      redistribute_input(t: f32[20, 10], trace: P->R)
        _c10d_functional::all_reduce(t: f32[20, 10], sum, 5)
  [node] hook_result_3: call_function[torch._dynamo.variables.tensor.prim_to_local](outputs_3)
  [node] output: output
```




cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

pianpwk committed:
Update on "[DebugMode][3/N] run_graph() method for fx.Node logging"

Allows logging on node calls, composed with existing `__torch_dispatch__` calls. Does this by creating a custom fx.Interpreter class with logging on `run_node()`.

Example from `debug_mode = DebugMode(); debug_mode.run_graph(gm, inputs)`:
```
  [node] l_flat_args_0_: placeholder
  [node] l__self____export_root_mlp_0_net1_weight: get_attr[L__self___mlp_0_net1_weight]()
  [node] l__self____export_root_mlp_0_net1_bias: get_attr[L__self___mlp_0_net1_bias]()
  [node] linear: call_function[torch._C._nn.linear](l_flat_args_0_, l__self____export_root_mlp_0_net1_weight, l__self____export_root_mlp_0_net1_bias)
    aten::t(dt: f32[16, 10]| S(0))
      aten::t(t: f32[4, 10])
    aten::addmm(dt: f32[16]| S(0), dt: f32[20, 10]| R, dt: f32[10, 16]| S(1))
      aten::addmm(t: f32[4], t: f32[20, 10], t: f32[10, 4])
  [node] outputs: call_function[torch._dynamo.variables.tensor.prim_redistribute](linear)
  [node] hook_result: call_function[torch._dynamo.variables.tensor.prim_to_local](outputs)
      aten::view(t: f32[20, 4], [20, 4])
  [node] input_tensor: call_function[torch.nn.functional.relu](hook_result, inplace=False)
      aten::relu(t: f32[20, 4])
  [node] input_tensor_1: call_function[torch._dynamo.variables.torch.prim from_local](input_tensor)
      aten::view(t: f32[20, 4], [20, 4])
  [node] l__self____export_root_mlp_0_net2_weight: get_attr[L__self___mlp_0_net2_weight]()
  [node] l__self____export_root_mlp_0_net2_bias: get_attr[L__self___mlp_0_net2_bias]()
  [node] linear_1: call_function[torch._C._nn.linear](input_tensor_1, l__self____export_root_mlp_0_net2_weight, l__self____export_root_mlp_0_net2_bias)
    aten::t(dt: f32[10, 16]| S(1))
      aten::t(t: f32[10, 4])
    aten::addmm(dt: f32[10]| R, dt: f32[20, 16]| S(1), dt: f32[16, 10]| S(0))
      redistribute_input(0, R -> P)
        redistribute_input(t: f32[10], trace: R->P)
          aten::div.Tensor(t: f32[10], 4)
      aten::addmm(t: f32[10], t: f32[20, 4], t: f32[4, 10])
  [node] outputs_1: call_function[torch._dynamo.variables.tensor.prim_redistribute](linear_1)
      redistribute_input(t: f32[20, 10], trace: P->R)
        _c10d_functional::all_reduce(t: f32[20, 10], sum, 5)
  [node] hook_result_1: call_function[torch._dynamo.variables.tensor.prim_to_local](outputs_1)
  [node] input_tensor_2: call_function[torch._dynamo.variables.torch.prim from_local](hook_result_1)
  [node] l__self____export_root_mlp_1_net1_weight: get_attr[L__self___mlp_1_net1_weight]()
  [node] l__self____export_root_mlp_1_net1_bias: get_attr[L__self___mlp_1_net1_bias]()
  [node] linear_2: call_function[torch._C._nn.linear](input_tensor_2, l__self____export_root_mlp_1_net1_weight, l__self____export_root_mlp_1_net1_bias)
    aten::t(dt: f32[16, 10]| S(0))
      aten::t(t: f32[4, 10])
    aten::addmm(dt: f32[16]| S(0), dt: f32[20, 10]| R, dt: f32[10, 16]| S(1))
  [node] outputs_2: call_function[torch._dynamo.variables.tensor.prim_redistribute](linear_2)
  [node] hook_result_2: call_function[torch._dynamo.variables.tensor.prim_to_local](outputs_2)
      aten::view(t: f32[20, 4], [20, 4])
  [node] input_tensor_3: call_function[torch.nn.functional.relu](hook_result_2, inplace=False)
      aten::relu(t: f32[20, 4])
  [node] input_tensor_4: call_function[torch._dynamo.variables.torch.prim from_local](input_tensor_3)
      aten::view(t: f32[20, 4], [20, 4])
  [node] l__self____export_root_mlp_1_net2_weight: get_attr[L__self___mlp_1_net2_weight]()
  [node] l__self____export_root_mlp_1_net2_bias: get_attr[L__self___mlp_1_net2_bias]()
  [node] linear_3: call_function[torch._C._nn.linear](input_tensor_4, l__self____export_root_mlp_1_net2_weight, l__self____export_root_mlp_1_net2_bias)
    aten::t(dt: f32[10, 16]| S(1))
      aten::t(t: f32[10, 4])
    aten::addmm(dt: f32[10]| R, dt: f32[20, 16]| S(1), dt: f32[16, 10]| S(0))
      redistribute_input(0, R -> P)
        redistribute_input(t: f32[10], trace: R->P)
          aten::div.Tensor(t: f32[10], 4)
      aten::addmm(t: f32[10], t: f32[20, 4], t: f32[4, 10])
  [node] outputs_3: call_function[torch._dynamo.variables.tensor.prim_redistribute](linear_3)
      redistribute_input(t: f32[20, 10], trace: P->R)
        _c10d_functional::all_reduce(t: f32[20, 10], sum, 5)
  [node] hook_result_3: call_function[torch._dynamo.variables.tensor.prim_to_local](outputs_3)
  [node] output: output
```




cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

pianpwk committed:
Update on "[DebugMode][3/N] run_graph() method for fx.Node logging"

Allows logging on node calls, composed with existing `__torch_dispatch__` calls. Does this by creating a custom fx.Interpreter class with logging on `run_node()`.

Example from `debug_mode = DebugMode(); debug_mode.run_graph(gm, inputs)`:
```
  [node] l_flat_args_0_: placeholder
  [node] l__self____export_root_mlp_0_net1_weight: get_attr[L__self___mlp_0_net1_weight]()
  [node] l__self____export_root_mlp_0_net1_bias: get_attr[L__self___mlp_0_net1_bias]()
  [node] linear: call_function[torch._C._nn.linear](l_flat_args_0_, l__self____export_root_mlp_0_net1_weight, l__self____export_root_mlp_0_net1_bias)
    aten::t(dt: f32[16, 10]| S(0))
      aten::t(t: f32[4, 10])
    aten::addmm(dt: f32[16]| S(0), dt: f32[20, 10]| R, dt: f32[10, 16]| S(1))
      aten::addmm(t: f32[4], t: f32[20, 10], t: f32[10, 4])
  [node] outputs: call_function[torch._dynamo.variables.tensor.prim_redistribute](linear)
  [node] hook_result: call_function[torch._dynamo.variables.tensor.prim_to_local](outputs)
      aten::view(t: f32[20, 4], [20, 4])
  [node] input_tensor: call_function[torch.nn.functional.relu](hook_result, inplace=False)
      aten::relu(t: f32[20, 4])
  [node] input_tensor_1: call_function[torch._dynamo.variables.torch.prim from_local](input_tensor)
      aten::view(t: f32[20, 4], [20, 4])
  [node] l__self____export_root_mlp_0_net2_weight: get_attr[L__self___mlp_0_net2_weight]()
  [node] l__self____export_root_mlp_0_net2_bias: get_attr[L__self___mlp_0_net2_bias]()
  [node] linear_1: call_function[torch._C._nn.linear](input_tensor_1, l__self____export_root_mlp_0_net2_weight, l__self____export_root_mlp_0_net2_bias)
    aten::t(dt: f32[10, 16]| S(1))
      aten::t(t: f32[10, 4])
    aten::addmm(dt: f32[10]| R, dt: f32[20, 16]| S(1), dt: f32[16, 10]| S(0))
      redistribute_input(0, R -> P)
        redistribute_input(t: f32[10], trace: R->P)
          aten::div.Tensor(t: f32[10], 4)
      aten::addmm(t: f32[10], t: f32[20, 4], t: f32[4, 10])
  [node] outputs_1: call_function[torch._dynamo.variables.tensor.prim_redistribute](linear_1)
      redistribute_input(t: f32[20, 10], trace: P->R)
        _c10d_functional::all_reduce(t: f32[20, 10], sum, 5)
  [node] hook_result_1: call_function[torch._dynamo.variables.tensor.prim_to_local](outputs_1)
  [node] input_tensor_2: call_function[torch._dynamo.variables.torch.prim from_local](hook_result_1)
  [node] l__self____export_root_mlp_1_net1_weight: get_attr[L__self___mlp_1_net1_weight]()
  [node] l__self____export_root_mlp_1_net1_bias: get_attr[L__self___mlp_1_net1_bias]()
  [node] linear_2: call_function[torch._C._nn.linear](input_tensor_2, l__self____export_root_mlp_1_net1_weight, l__self____export_root_mlp_1_net1_bias)
    aten::t(dt: f32[16, 10]| S(0))
      aten::t(t: f32[4, 10])
    aten::addmm(dt: f32[16]| S(0), dt: f32[20, 10]| R, dt: f32[10, 16]| S(1))
  [node] outputs_2: call_function[torch._dynamo.variables.tensor.prim_redistribute](linear_2)
  [node] hook_result_2: call_function[torch._dynamo.variables.tensor.prim_to_local](outputs_2)
      aten::view(t: f32[20, 4], [20, 4])
  [node] input_tensor_3: call_function[torch.nn.functional.relu](hook_result_2, inplace=False)
      aten::relu(t: f32[20, 4])
  [node] input_tensor_4: call_function[torch._dynamo.variables.torch.prim from_local](input_tensor_3)
      aten::view(t: f32[20, 4], [20, 4])
  [node] l__self____export_root_mlp_1_net2_weight: get_attr[L__self___mlp_1_net2_weight]()
  [node] l__self____export_root_mlp_1_net2_bias: get_attr[L__self___mlp_1_net2_bias]()
  [node] linear_3: call_function[torch._C._nn.linear](input_tensor_4, l__self____export_root_mlp_1_net2_weight, l__self____export_root_mlp_1_net2_bias)
    aten::t(dt: f32[10, 16]| S(1))
      aten::t(t: f32[10, 4])
    aten::addmm(dt: f32[10]| R, dt: f32[20, 16]| S(1), dt: f32[16, 10]| S(0))
      redistribute_input(0, R -> P)
        redistribute_input(t: f32[10], trace: R->P)
          aten::div.Tensor(t: f32[10], 4)
      aten::addmm(t: f32[10], t: f32[20, 4], t: f32[4, 10])
  [node] outputs_3: call_function[torch._dynamo.variables.tensor.prim_redistribute](linear_3)
      redistribute_input(t: f32[20, 10], trace: P->R)
        _c10d_functional::all_reduce(t: f32[20, 10], sum, 5)
  [node] hook_result_3: call_function[torch._dynamo.variables.tensor.prim_to_local](outputs_3)
  [node] output: output
```




cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

pianpwk committed:
Update on "[DebugMode][3/N] run_graph() method for fx.Node logging"

Allows logging on node calls, composed with existing `__torch_dispatch__` calls. Does this by creating a custom fx.Interpreter class with logging on `run_node()`.

Example from `debug_mode = DebugMode(); debug_mode.run_graph(gm, inputs)`:
```
  [node] l_flat_args_0_: placeholder
  [node] l__self____export_root_mlp_0_net1_weight: get_attr[L__self___mlp_0_net1_weight]()
  [node] l__self____export_root_mlp_0_net1_bias: get_attr[L__self___mlp_0_net1_bias]()
  [node] linear: call_function[torch._C._nn.linear](l_flat_args_0_, l__self____export_root_mlp_0_net1_weight, l__self____export_root_mlp_0_net1_bias)
    aten::t(dt: f32[16, 10]| S(0))
      aten::t(t: f32[4, 10])
    aten::addmm(dt: f32[16]| S(0), dt: f32[20, 10]| R, dt: f32[10, 16]| S(1))
      aten::addmm(t: f32[4], t: f32[20, 10], t: f32[10, 4])
  [node] outputs: call_function[torch._dynamo.variables.tensor.prim_redistribute](linear)
  [node] hook_result: call_function[torch._dynamo.variables.tensor.prim_to_local](outputs)
      aten::view(t: f32[20, 4], [20, 4])
  [node] input_tensor: call_function[torch.nn.functional.relu](hook_result, inplace=False)
      aten::relu(t: f32[20, 4])
  [node] input_tensor_1: call_function[torch._dynamo.variables.torch.prim from_local](input_tensor)
      aten::view(t: f32[20, 4], [20, 4])
  [node] l__self____export_root_mlp_0_net2_weight: get_attr[L__self___mlp_0_net2_weight]()
  [node] l__self____export_root_mlp_0_net2_bias: get_attr[L__self___mlp_0_net2_bias]()
  [node] linear_1: call_function[torch._C._nn.linear](input_tensor_1, l__self____export_root_mlp_0_net2_weight, l__self____export_root_mlp_0_net2_bias)
    aten::t(dt: f32[10, 16]| S(1))
      aten::t(t: f32[10, 4])
    aten::addmm(dt: f32[10]| R, dt: f32[20, 16]| S(1), dt: f32[16, 10]| S(0))
      redistribute_input(0, R -> P)
        redistribute_input(t: f32[10], trace: R->P)
          aten::div.Tensor(t: f32[10], 4)
      aten::addmm(t: f32[10], t: f32[20, 4], t: f32[4, 10])
  [node] outputs_1: call_function[torch._dynamo.variables.tensor.prim_redistribute](linear_1)
      redistribute_input(t: f32[20, 10], trace: P->R)
        _c10d_functional::all_reduce(t: f32[20, 10], sum, 5)
  [node] hook_result_1: call_function[torch._dynamo.variables.tensor.prim_to_local](outputs_1)
  [node] input_tensor_2: call_function[torch._dynamo.variables.torch.prim from_local](hook_result_1)
  [node] l__self____export_root_mlp_1_net1_weight: get_attr[L__self___mlp_1_net1_weight]()
  [node] l__self____export_root_mlp_1_net1_bias: get_attr[L__self___mlp_1_net1_bias]()
  [node] linear_2: call_function[torch._C._nn.linear](input_tensor_2, l__self____export_root_mlp_1_net1_weight, l__self____export_root_mlp_1_net1_bias)
    aten::t(dt: f32[16, 10]| S(0))
      aten::t(t: f32[4, 10])
    aten::addmm(dt: f32[16]| S(0), dt: f32[20, 10]| R, dt: f32[10, 16]| S(1))
  [node] outputs_2: call_function[torch._dynamo.variables.tensor.prim_redistribute](linear_2)
  [node] hook_result_2: call_function[torch._dynamo.variables.tensor.prim_to_local](outputs_2)
      aten::view(t: f32[20, 4], [20, 4])
  [node] input_tensor_3: call_function[torch.nn.functional.relu](hook_result_2, inplace=False)
      aten::relu(t: f32[20, 4])
  [node] input_tensor_4: call_function[torch._dynamo.variables.torch.prim from_local](input_tensor_3)
      aten::view(t: f32[20, 4], [20, 4])
  [node] l__self____export_root_mlp_1_net2_weight: get_attr[L__self___mlp_1_net2_weight]()
  [node] l__self____export_root_mlp_1_net2_bias: get_attr[L__self___mlp_1_net2_bias]()
  [node] linear_3: call_function[torch._C._nn.linear](input_tensor_4, l__self____export_root_mlp_1_net2_weight, l__self____export_root_mlp_1_net2_bias)
    aten::t(dt: f32[10, 16]| S(1))
      aten::t(t: f32[10, 4])
    aten::addmm(dt: f32[10]| R, dt: f32[20, 16]| S(1), dt: f32[16, 10]| S(0))
      redistribute_input(0, R -> P)
        redistribute_input(t: f32[10], trace: R->P)
          aten::div.Tensor(t: f32[10], 4)
      aten::addmm(t: f32[10], t: f32[20, 4], t: f32[4, 10])
  [node] outputs_3: call_function[torch._dynamo.variables.tensor.prim_redistribute](linear_3)
      redistribute_input(t: f32[20, 10], trace: P->R)
        _c10d_functional::all_reduce(t: f32[20, 10], sum, 5)
  [node] hook_result_3: call_function[torch._dynamo.variables.tensor.prim_to_local](outputs_3)
  [node] output: output
```




cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

pianpwk committed:
Update on "[DebugMode][3/N] run_graph() method for fx.Node logging"

Allows logging on node calls, composed with existing `__torch_dispatch__` calls. Does this by creating a custom fx.Interpreter class with logging on `run_node()`.

Example from `debug_mode = DebugMode(); debug_mode.run_graph(gm, inputs)`:
```
  [node] l_flat_args_0_: placeholder
  [node] l__self____export_root_mlp_0_net1_weight: get_attr[L__self___mlp_0_net1_weight]()
  [node] l__self____export_root_mlp_0_net1_bias: get_attr[L__self___mlp_0_net1_bias]()
  [node] linear: call_function[torch._C._nn.linear](l_flat_args_0_, l__self____export_root_mlp_0_net1_weight, l__self____export_root_mlp_0_net1_bias)
    aten::t(dt: f32[16, 10]| S(0))
      aten::t(t: f32[4, 10])
    aten::addmm(dt: f32[16]| S(0), dt: f32[20, 10]| R, dt: f32[10, 16]| S(1))
      aten::addmm(t: f32[4], t: f32[20, 10], t: f32[10, 4])
  [node] outputs: call_function[torch._dynamo.variables.tensor.prim_redistribute](linear)
  [node] hook_result: call_function[torch._dynamo.variables.tensor.prim_to_local](outputs)
      aten::view(t: f32[20, 4], [20, 4])
  [node] input_tensor: call_function[torch.nn.functional.relu](hook_result, inplace=False)
      aten::relu(t: f32[20, 4])
  [node] input_tensor_1: call_function[torch._dynamo.variables.torch.prim from_local](input_tensor)
      aten::view(t: f32[20, 4], [20, 4])
  [node] l__self____export_root_mlp_0_net2_weight: get_attr[L__self___mlp_0_net2_weight]()
  [node] l__self____export_root_mlp_0_net2_bias: get_attr[L__self___mlp_0_net2_bias]()
  [node] linear_1: call_function[torch._C._nn.linear](input_tensor_1, l__self____export_root_mlp_0_net2_weight, l__self____export_root_mlp_0_net2_bias)
    aten::t(dt: f32[10, 16]| S(1))
      aten::t(t: f32[10, 4])
    aten::addmm(dt: f32[10]| R, dt: f32[20, 16]| S(1), dt: f32[16, 10]| S(0))
      redistribute_input(0, R -> P)
        redistribute_input(t: f32[10], trace: R->P)
          aten::div.Tensor(t: f32[10], 4)
      aten::addmm(t: f32[10], t: f32[20, 4], t: f32[4, 10])
  [node] outputs_1: call_function[torch._dynamo.variables.tensor.prim_redistribute](linear_1)
      redistribute_input(t: f32[20, 10], trace: P->R)
        _c10d_functional::all_reduce(t: f32[20, 10], sum, 5)
  [node] hook_result_1: call_function[torch._dynamo.variables.tensor.prim_to_local](outputs_1)
  [node] input_tensor_2: call_function[torch._dynamo.variables.torch.prim from_local](hook_result_1)
  [node] l__self____export_root_mlp_1_net1_weight: get_attr[L__self___mlp_1_net1_weight]()
  [node] l__self____export_root_mlp_1_net1_bias: get_attr[L__self___mlp_1_net1_bias]()
  [node] linear_2: call_function[torch._C._nn.linear](input_tensor_2, l__self____export_root_mlp_1_net1_weight, l__self____export_root_mlp_1_net1_bias)
    aten::t(dt: f32[16, 10]| S(0))
      aten::t(t: f32[4, 10])
    aten::addmm(dt: f32[16]| S(0), dt: f32[20, 10]| R, dt: f32[10, 16]| S(1))
  [node] outputs_2: call_function[torch._dynamo.variables.tensor.prim_redistribute](linear_2)
  [node] hook_result_2: call_function[torch._dynamo.variables.tensor.prim_to_local](outputs_2)
      aten::view(t: f32[20, 4], [20, 4])
  [node] input_tensor_3: call_function[torch.nn.functional.relu](hook_result_2, inplace=False)
      aten::relu(t: f32[20, 4])
  [node] input_tensor_4: call_function[torch._dynamo.variables.torch.prim from_local](input_tensor_3)
      aten::view(t: f32[20, 4], [20, 4])
  [node] l__self____export_root_mlp_1_net2_weight: get_attr[L__self___mlp_1_net2_weight]()
  [node] l__self____export_root_mlp_1_net2_bias: get_attr[L__self___mlp_1_net2_bias]()
  [node] linear_3: call_function[torch._C._nn.linear](input_tensor_4, l__self____export_root_mlp_1_net2_weight, l__self____export_root_mlp_1_net2_bias)
    aten::t(dt: f32[10, 16]| S(1))
      aten::t(t: f32[10, 4])
    aten::addmm(dt: f32[10]| R, dt: f32[20, 16]| S(1), dt: f32[16, 10]| S(0))
      redistribute_input(0, R -> P)
        redistribute_input(t: f32[10], trace: R->P)
          aten::div.Tensor(t: f32[10], 4)
      aten::addmm(t: f32[10], t: f32[20, 4], t: f32[4, 10])
  [node] outputs_3: call_function[torch._dynamo.variables.tensor.prim_redistribute](linear_3)
      redistribute_input(t: f32[20, 10], trace: P->R)
        _c10d_functional::all_reduce(t: f32[20, 10], sum, 5)
  [node] hook_result_3: call_function[torch._dynamo.variables.tensor.prim_to_local](outputs_3)
  [node] output: output
```




cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]
